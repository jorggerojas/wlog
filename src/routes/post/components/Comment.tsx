// @flow
import React from "react";
import cookie from "react-cookies";
import { Link } from "react-router-dom";
import { parseDate } from "../../../config";

interface CommentProps {
  comment: string;
  user: string;
  date: string;
  id: string;
}

const Comment = ({ comment, user, date, id }: CommentProps) => {
  return (
    <article className="uk-comment uk-comment-primary uk-margin">
      <header
        className="uk-comment-header uk-grid-medium uk-flex-middle"
        uk-grid=""
      >
        <div className="uk-width-expand">
          <p
            style={{ fontSize: "1rem" }}
            className="uk-article-title uk-button uk-button-text"
          >
            <Link className="uk-link-reset" to={`/user/${user}`}>
              {user}
            </Link>
          </p>
          {cookie.load("USER") === user ? (
            <div className="uk-align-right">
              <span
                style={{ cursor: "pointer" }}
                uk-icon="icon:close"
                onClick={() => console.log(id)}
              />
            </div>
          ) : null}
          <ul className="uk-comment-meta uk-subnav uk-text-left uk-margin-remove-top">
            <li className="uk-text-italic uk-padding-remove">
              {parseDate(date)}
            </li>
          </ul>
        </div>
      </header>
      <div className="uk-comment-body">
        <p className="uk-text-italic">{`${comment
          .toUpperCase()
          .substr(0, 1)}${comment
          .toLowerCase()
          .substr(1, comment.length - 1)}`}</p>
      </div>
    </article>
  );
};

export default Comment;
