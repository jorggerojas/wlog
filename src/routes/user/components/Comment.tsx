// @flow
import React from "react";
import { Link } from "react-router-dom";

interface CommentProps {
  comment: string;
  date: string;
  post: string;
  id: string;
}

const Comment = ({ comment, date, post, id }: CommentProps) => {
  return (
    <div className="uk-card-small uk-card-default uk-card-hover">
      <div className="uk-card-body">
        <span className="uk-text-meta uk-text-italic">{date}</span>
        <p className="uk-text-secondary uk-text-italic">"{comment}"</p>
      </div>
      <div className="uk-card-footer uk-text-right">
        <Link to={`/post/${id}`} className="uk-link-heading">
          Ver Post{" "}
        </Link>
      </div>
    </div>
  );
};

export default Comment;
