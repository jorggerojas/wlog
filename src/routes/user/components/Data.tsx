// @flow
import React from "react";
import swal from "sweetalert";
import { removeCookie, URL } from "../../../config";
import axios from "axios";
import cookie from "react-cookies";

interface DataProps {
  username: string;
  role: string;
  date: string;
  myAccount: boolean;
  admin: boolean;
  loading: boolean;
  setLoading: Function;
  blocked: string;
  setBlocked: Function;
}

const Data = ({
  username,
  role,
  date,
  myAccount,
  admin,
  loading,
  blocked,
  setBlocked,
  setLoading,
}: DataProps) => {
  const logOut = () => {
    swal({
      title: "Are you sure want to log out :(?",
      buttons: {
        cancel: {
          visible: true,
          text: "No",
        },
        ok: {
          visible: true,
          text: "Yes",
        },
      },
    }).then((willLogOut: boolean) => {
      if (willLogOut) {
        setLoading(true);
        setTimeout(() => {
          removeCookie("USER");
          removeCookie("ROLE");
          removeCookie("TOKEN");
          setTimeout(() => {
            setLoading(false);
            window.location.href = "/";
          }, 100);
        }, 500);
      } else return;
    });
  };
  const deleteAccount = () => {
    swal({
      title: "Are you sure want to delete your account?",
      text:
        "All your data, posts, comments and related information will be destroyed forever.",
      icon: "warning",
      buttons: {
        cancel: {
          visible: true,
          text: "No",
        },
        ok: {
          visible: true,
          text: "Yes",
        },
      },
      dangerMode: true,
    }).then((willDelete: boolean) => {
      if (willDelete) {
        setLoading(!loading);
        setTimeout(() => {
          deleteForeverAndEver();
        }, 500);
      } else return;
    });
  };
  const deleteForeverAndEver = () => {
    axios({
      method: "delete",
      url: `${URL}/users/${username}`,
      headers: {
        Authorization: `Bearer ${cookie.load("TOKEN")}`,
      },
    })
      .then(() => {
        removeCookie("USER");
        removeCookie("ROLE");
        removeCookie("TOKEN");
        setTimeout(() => {
          swal("Poof! Your account and all the data have been deleted!", {
            icon: "success",
          }).then(() => {
            setLoading(false);
            setTimeout(() => {
              window.location.href = "/";
            }, 100);
          });
        }, 500);
      })
      .catch((error) => {
        if (error.response.status !== 404) {
          swal("You can't make this move your role", {
            icon: "error",
          });
        } else {
          swal(
            "We can't delete the account because it's missing or not exist",
            {
              icon: "error",
            }
          );
        }
        setLoading(false);
      });
  };
  const enable = (isBlocked: string) => {
    setLoading(true);
    axios({
      method: "put",
      url: `${URL}/users/${username}`,
      headers: {
        Authorization: `Bearer ${cookie.load("TOKEN")}`,
      },
      data: {
        isBlocked: `${isBlocked === "1" ? "0" : "1"}`,
      },
    })
      .then(() => {
        setBlocked(isBlocked === "1" ? "0" : "1");
        setTimeout(() => {
          setLoading(false);
          isBlocked === "1"
            ? swal({
                title: "Account enabled",
              })
            : swal({
                title: "Account disabled",
              });
        }, 500);
      })
      .catch((error: any) => {
        if (error.response.status === 404) {
          swal(
            "We can't change the account because it's missing or not exist",
            {
              icon: "error",
            }
          );
        } else {
          swal("You can't make this move your role", {
            icon: "error",
          });
        }
      });
  };
  return (
    <div className={`uk-text-center`}>
      <h2
        className="uk-text-bolder uk-text-truncate"
        title={username.toUpperCase()}
        style={{
          fontSize: username.length < 11 ? "2.5rem" : `1.9rem`,
        }}
      >
        {username.toUpperCase()}
      </h2>
      <div className="uk-margin">
        <p className="uk-margin-remove-top uk-text-meta uk-text-italic">
          Member since: {date}
        </p>
        {blocked === "1" ? (
          <p className="uk-margin-remove-top uk-text-secondary uk-text-bolder">
            BLOCKED
          </p>
        ) : null}
        <p className="uk-text-bold">{role.toUpperCase()}</p>
        <div className="uk-grid uk-flex-around" uk-grid="">
          {myAccount ? (
            <div className="uk-width-1-1 uk-margin-small">
              <button
                className="uk-button-text uk-button uk-text-danger uk-margin"
                onClick={logOut}
              >
                Log out
              </button>
            </div>
          ) : null}
          {admin && !myAccount ? (
            <div>
              <button
                className="uk-button uk-button-text uk-text-danger uk-margin"
                onClick={() => enable(blocked)}
              >
                {blocked === "1" ? "Enable account" : "Disable account"}
              </button>
            </div>
          ) : null}
          {admin || myAccount ? (
            <div>
              <button
                className="uk-button uk-button-danger uk-margin"
                onClick={deleteAccount}
              >
                Delete account
              </button>
            </div>
          ) : null}
        </div>
      </div>
    </div>
  );
};

export default Data;
