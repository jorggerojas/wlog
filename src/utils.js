import React from "react";
import { render, waitFor } from "@testing-library/react";
import { createMemoryHistory } from "history";
import { Router, Route } from "react-router-dom";
import "@testing-library/jest-dom";

const renderWithRouter = (ui, { route = "/", ...renderOptions } = {}) => {
  const history = createMemoryHistory({ initialEntries: [route] });
  const utils = render(<Router history={history}>{ui}</Router>, renderOptions);
  const finishLoading = () =>
    waitFor(() =>
      expect(utils.queryByText("Loading, please wait...")).toBeNull()
    );
  return {
    ...utils,
    finishLoading,
    history,
  };
};

export {
  Simulate,
  wait,
  render,
  cleanup,
  renderIntoDocument,
  fireEvent,
} from "@testing-library/react";
export { renderWithRouter };

export function renderWithProviders(
  ui,
  {
    route = "/",
    history = createMemoryHistory({ initialEntries: [route] }),
  } = {},
  apolloMocks
) {
  console.log("I am route inside renderWithProviders Wrapper:", route);
  return {
    ...render(
      <Router history={history}>
        <Route path={route}>{ui}</Route>
      </Router>
    ),
    history,
  };
}
