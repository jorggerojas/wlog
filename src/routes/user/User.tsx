import React, { useEffect, useState } from "react";
import t from "typy";
import ReactPaginate from "react-paginate";
import { useParams } from "react-router-dom";
import axios from "axios";
import Badge from "../main/components/Badge";
import Header from "../main/components/Header";
import PostMinified from "../post/components/PostMinified";
import Data from "./components/Data";
import Comment from "./components/Comment";
import NoMatch from "../NoMatch";
import { loadStorage, parseDate, URL } from "../../config";
import Loading from "../loading/Loading";
import { Container, DivComments } from "../../styles/containers";
import { Title4 } from "../../styles/text";

interface UserProps {
  theme: boolean;
  handle: Function;
}

const User = ({ theme, handle }: UserProps) => {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const [__, setUserData] = useState();
  const [loading, setLoading] = useState(false);
  const [pageCountPost, setPageCountPost] = useState(1);
  const [pageCountComment, setPageCountComment] = useState(1);
  const [posts, setPosts] = useState([]);
  const [returnMatch, setReturnMatch] = useState(false);
  const [comments, setComments] = useState([]);
  const USER = loadStorage("USER");
  const ROLE = loadStorage("ROLE");
  const [role, setRole] = useState("...");
  const [notFoundPosts, setPostsNotFound] = useState(false);
  const [notFoundComments, setCommentsNotFound] = useState(false);
  const [blocked, setBlocked] = useState("0");
  const [date, setDate] = useState("1998-07-19");
  const { username } = useParams<{ username: string }>();
  const getComments = (page?: number, size?: number) => {
    return axios.get(
      `${URL}/users/${username}/posts/all/comments?page=${page ?? 0}&size=${
        size ?? 5
      }`
    );
  };
  const changeContentComments = ({ selected }: any) => {
    getComments(selected)
      .then(({ data }: any) => {
        setComments(data.content);
      })
      .catch(() => {
        setComments([]);
        setCommentsNotFound(!notFoundComments);
      });
  };
  const getUserData = () => {
    return axios.get(`${URL}/users/${username}`);
  };
  const getPosts = () => {
    return axios.get(`${URL}/users/${username}/posts?size=4`);
  };
  useEffect(() => {
    getUserData()
      .then(({ data }: any) => {
        setRole(data.role[0]);
        setDate(parseDate(data.dateLog));
        setUserData(data);
        setBlocked(data.isBlocked);
        getPosts()
          .then((response: any) => {
            setPageCountPost(Math.ceil(response.data.totalElements / 4));
            setPosts(response.data.content);
            if (ROLE === "ADMIN" || username === USER) {
              getComments()
                .then(({ data }: any) => {
                  setPageCountComment(
                    Math.ceil(response.data.totalElements / 3)
                  );
                  setComments(data.content);
                })
                .catch(() => {
                  setComments([]);
                  setCommentsNotFound(!notFoundComments);
                });
            }
          })
          .catch(() => {
            setPosts([]);
            setPostsNotFound(!notFoundPosts);
          });
      })
      .catch(() => {
        setReturnMatch(true);
      });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [username]);
  const changeContent = ({ selected }: any) => {
    axios
      .get(`${URL}/users/${username}/posts?page=${selected}&size=4`)
      .then(({ data }: any) => {
        setPosts(data.content);
      });
  };
  const data = [
    { title: "POSTS", isActive: true },
    { title: "COMMENTS", isActive: false },
  ];
  if (t(username).isUndefined) return <NoMatch theme={theme} handle={handle} />;
  if (t(returnMatch).isTrue) return <NoMatch theme={theme} handle={handle} />;
  else {
    if ((ROLE === "LECTOR" || t(ROLE).isNullOrUndefined) && blocked === "1") {
      return <NoMatch theme={theme} handle={handle} />;
    }
    if (t(loading).isTrue) {
      return <Loading load={loading} />;
    }
    return (
      <Container
        className={`uk-animation-fade ${t(loading).isTrue ? "blur" : ""}`}
      >
        <Header theme={theme} handle={handle} />
        <div
          className="uk-flex-top uk-flex-center@s uk-padding-small"
          uk-grid=""
        >
          <div className="uk-width-2-3@m">
            <ul
              className="uk-subnav uk-subnav-pill uk-margin-remove-top"
              uk-switcher=""
            >
              <Badge
                link={true}
                key={data[0].title}
                title={data[0].title}
                isActive={data[0].isActive}
                target={"0"}
              />
              {role === "ADMIN" || username === USER ? (
                <Badge
                  link={true}
                  key={data[1].title}
                  title={data[1].title}
                  isActive={data[1].isActive}
                  target={"1"}
                />
              ) : null}
            </ul>
            <ul className="uk-switcher uk-margin">
              <li className="uk-active">
                <DivComments
                  className="uk-grid uk-grid-match uk-child-width-1-2@s uk-child-width-1-2@m "
                  uk-grid=""
                >
                  {!t(posts).isNull && posts.length ? (
                    posts.map((post: any) => (
                      <div className="uk-margin-bottom match" key={post.index}>
                        <PostMinified
                          title={post.title}
                          date={parseDate(post.dateLog)}
                          summary={post.summary}
                          badge={post.keywords[0]}
                          id={post.index}
                        />
                      </div>
                    ))
                  ) : (
                    <div className=" uk-margin">
                      <Title4 className="uk-text-italic">
                        No posts available
                      </Title4>
                    </div>
                  )}
                </DivComments>
                {!t(posts).isNull && posts.length ? (
                  <div className="paginate">
                    <ReactPaginate
                      pageCount={pageCountPost}
                      onPageChange={changeContent}
                      pageRangeDisplayed={2}
                      marginPagesDisplayed={4}
                      breakLabel="..."
                      activeClassName="uk-active active"
                      activeLinkClassName="uk-active active"
                      disabledClassName="uk-disabled"
                      nextClassName="uk-pagination-next"
                      previousClassName="uk-pagination-previous"
                      pageClassName=""
                      containerClassName="uk-pagination uk-flex-center"
                      nextLabel="Next"
                    />
                  </div>
                ) : null}
              </li>
              {ROLE === "ADMIN" || username === USER ? (
                <li>
                  <DivComments
                    style={{ minHeight: "68vh" }}
                    className="uk-grid uk-margin uk-child-width-1-2@s uk-child-width-1-2@m uk-child-width-1-3@l uk-child-width-1-4@xl"
                    uk-grid=""
                  >
                    {!t(comments).isNull && comments.length ? (
                      comments.map((comment: any) => (
                        <div
                          className="uk-margin-small-bottom"
                          key={comment.index}
                        >
                          <Comment
                            comment={comment.content}
                            date={parseDate(comment.dateLog)}
                            post={comment.post}
                            user={USER ?? ""}
                          />
                        </div>
                      ))
                    ) : (
                      <div className="uk-text-center uk-margin">
                        <Title4 className="">No comments available</Title4>
                      </div>
                    )}
                  </DivComments>
                  {!t(comments).isNull && comments.length ? (
                    <div className="paginate">
                      <ReactPaginate
                        pageCount={pageCountComment}
                        onPageChange={changeContentComments}
                        pageRangeDisplayed={2}
                        marginPagesDisplayed={4}
                        breakLabel="..."
                        activeClassName="uk-active active"
                        activeLinkClassName="uk-active active"
                        disabledClassName="uk-disabled"
                        nextClassName="uk-pagination-next"
                        previousClassName="uk-pagination-previous"
                        pageClassName=""
                        containerClassName="uk-pagination uk-flex-center"
                        nextLabel="Next"
                      />
                    </div>
                  ) : null}
                </li>
              ) : null}
            </ul>
          </div>
          <div className="uk-width-1-3@m uk-flex-first">
            <div uk-sticky="media: @m; top: 15vh; offset:15vh; animation: uk-animation-slide-top-small">
              <Data
                username={username}
                date={date}
                loading={loading}
                setLoading={setLoading}
                role={role}
                blocked={blocked}
                setBlocked={setBlocked}
                myAccount={username === USER}
                admin={ROLE === "ADMIN"}
              />
            </div>
          </div>
        </div>
      </Container>
    );
  }
};

export default User;
