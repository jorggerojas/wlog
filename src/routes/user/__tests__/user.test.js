import "@testing-library/jest-dom/extend-expect";
import React from "react";
import ReactDOM from "react-dom";
import { Route, BrowserRouter as Router } from "react-router-dom";
import User from "../User";
import Data from "../components/Data";
import Comment from "../components/Comment";
import { renderWithProviders } from "../../../utils";
import { shallow, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { act } from "@testing-library/react";

configure({ adapter: new Adapter() });

let container;
describe("<User/> tree component", () => {
  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
  });
  afterEach(() => {
    document.body.removeChild(container);
    container = null;
  });
  test("Render <User/> component with user mamberroi", async () => {
    const { findByText } = renderWithProviders(
      <Route path="/user/:username">
        <User />
      </Route>,
      {
        route: "/user/mamberroi",
      }
    );
    await findByText("POSTS");
  });
  test("Verify <Data/> component in <User/>", async () => {
    const dataProps = {
      username: "mamberroi",
    };
    const { findByText } = renderWithProviders(
      <Route path="/user/:username">
        <User>
          <Data {...dataProps}></Data>
        </User>
      </Route>,
      {
        route: `/user/${dataProps.username}`,
      }
    );
    await findByText(dataProps.username.toUpperCase());
  });
  test("Verify <Comment/> render the props", async () => {
    const commentProps = {
      comment: "This is a comment",
      user: "mamberroi",
      date: "2020-12-12",
      id: "1",
      post: "2",
    };
    const wrapper = shallow(<Comment {...commentProps}></Comment>);
    await wrapper.containsMatchingElement(commentProps.comment);
  });
});
