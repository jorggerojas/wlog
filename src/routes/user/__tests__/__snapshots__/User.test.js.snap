// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<User/> tree component <Comment/> component render component with props 1`] = `
<div>
  <div
    class="uk-card-small uk-card-default uk-card-hover"
  >
    <div
      class="uk-card-body"
    >
      <span
        class="uk-text-meta uk-text-italic"
      >
        2020-12-12
      </span>
      <p
        class="uk-text-secondary uk-text-italic"
      >
        "
        This is a comment
        "
      </p>
    </div>
    <div
      class="uk-card-footer uk-text-right"
    >
      <a
        class="uk-link-heading"
        href="/user/mamberroi/post/2"
      >
        Ver Post
         
      </a>
    </div>
  </div>
</div>
`;

exports[`<User/> tree component <Data/> component render component with props 1`] = `
<div>
  <div
    class="uk-text-center"
  >
    <h2
      class="sc-iktFzd clEUXB uk-text-bolder uk-text-truncate fontSizeLarge"
      title="REALUSER"
    >
      REALUSER
    </h2>
    <div
      class="uk-margin"
    >
      <p
        class="uk-margin-remove-top uk-text-meta uk-text-italic"
      >
        Member since: 
      </p>
      <p
        class="uk-text-bold"
      >
        ADMIN
      </p>
      <div
        class="uk-grid uk-flex-around"
        uk-grid=""
      />
    </div>
  </div>
</div>
`;

exports[`<User/> tree component render <User/> component with invalid username 1`] = `
<div>
  <div
    class="sc-gsTCUz jjwlfv uk-animation-fade uk-padding-large"
  >
    <h3
      class="sc-bkzZxe cNxjiI"
    >
      No match for 
      <code
        class="sc-idOhPF dMTeBE"
      >
        /user/realUser
      </code>
    </h3>
    <br />
    <a
      href="/"
    >
      <code
        class="sc-idOhPF dMTeBE"
      >
        Return to the main page
      </code>
    </a>
  </div>
</div>
`;
